pkg/controller/groups/samlgrouplinks/controller.go

```diff
@@ -0,0 +1,189 @@
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane-contrib/provider-gitlab/apis/groups/v1alpha1"
+	secretstoreapi "github.com/crossplane-contrib/provider-gitlab/apis/v1alpha1"
+	"github.com/crossplane-contrib/provider-gitlab/pkg/clients"
+	"github.com/crossplane-contrib/provider-gitlab/pkg/clients/groups"
+	"github.com/crossplane-contrib/provider-gitlab/pkg/features"
+)
+const (
+	errNotSamlGroupLink = "managed resource is not a SamlGroupLink custom resource"
+	errGetFailed            = "cannot get Gitlab SamlGroupLink"
+	errCreateFailed         = "cannot create Gitlab SamlGroupLink"
+	errDeleteFailed         = "cannot delete Gitlab SamlGroupLink"
+	errSamlGroupLinktNotFound = "cannot find Gitlab SamlGroupLink"
+	errMissingGroupID       = "missing Spec.ForProvider.GroupID"
+)
+
+func SetupSamlGroupLink(mgr ctrl.Manager, o controller.Options) error {
+	name := managed.ControllerName(v1alpha1.SamlGroupLinkKind)
+	cps := []managed.ConnectionPublisher{managed.NewAPISecretPublisher(mgr.GetClient(), mgr.GetScheme())}
+
+	if o.Features.Enabled(features.EnableAlphaExternalSecretStores) {
+		cps = append(cps, connection.NewDetailsManager(mgr.GetClient(), secretstoreapi.StoreConfigGroupVersionKind))
+	}
+
+	reconcilerOpts := []managed.ReconcilerOption{
+		managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), newGitlabClientFn: groups.NewSamlGroupLinkClient}),
+		managed.WithInitializers(),
+		managed.WithPollInterval(o.PollInterval),
+		managed.WithLogger(o.Logger.WithValues("controller", name)),
+		managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))),
+		managed.WithConnectionPublishers(cps...),
+	}
+	if o.Features.Enabled(features.EnableAlphaManagementPolicies) {
+		reconcilerOpts = append(reconcilerOpts, managed.WithManagementPolicies())
+	}
+
+	r := managed.NewReconciler(mgr, 
+		resource.ManagedKind(v1alpha1.SamlGroupLinkGroupVersionKind),
+		reconcilerOpts...)
+	
+		return ctrl.NewControllerManagedBy(mgr).
+			Named(name).
+			For(&v1alpha1.SamlGroupLink{}).
+			Complete(r)
+}
+
+type connector struct {
+	kube client.Client
+	newGitlabClientFn func(cfg clients.Config) groups.SamlGroupLinkClient
+}
+
+func (c * connector)  Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1alpha1.SamlGroupLink)
+	if !ok {
+		return nil, errors.New(errNotSamlGroupLink)
+	}
+
+	cfg, err := clients.GetConfig(ctx, c.kube, cr)
+	if err != nil {
+		return nil, err 
+	}
+	return &external{kube: c.kube, client: c.newGitlabClientFn(*cfg)}, nil
+}
+
+type external struct {
+	kube client.Client
+	client groups.SamlGroupLinkClient
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.SamlGroupLink)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotSamlGroupLink)
+	}
+
+	samlGroupName := *cr.Spec.ForProvider.Name
```

Comment:
External resource names should be stored in the external name annotation are retrieved with `meta.GetExternalName(cr)`.

Reply:
So for this 'samlGroupName' variable, its not trying to get the name of the object but the samlGroupName that gets passed to Gitlab. My idea behind having that samlGroupName explicitly defined in the forProvider block, is if users want to assign the same saml group to multiple gitlab groups. Let me know if you think this makes sense! 

e.g:
``` yaml
apiVersion: groups.gitlab.crossplane.io/v1alpha1
kind: SamlGroupLink
metadata:
  name: mysamlgroup-group1
spec:
  forProvider:
    groupIdRef:
      name: gitlab-group-1
    name: MySamlGroup
---
apiVersion: groups.gitlab.crossplane.io/v1alpha1
kind: SamlGroupLink
metadata:
  name: mysamlgroup-group2
spec:
  forProvider:
    groupIdRef:
      name: gitlab-group-2
    name: MySamlGroup
```


I could also change the json field to be more explicit that it is specific to the samlgroup as maybe 'name' is too generic. The reason I left it as 'name' is to be consistent with the upstream gitlab sdk. 

Could look something like e.g
``` yaml
apiVersion: groups.gitlab.crossplane.io/v1alpha1
kind: SamlGroupLink
metadata:
  name: mysamlgroup-group2
spec:
  forProvider:
    groupIdRef:
      name: gitlab-group-2
    samlGroupName: MySamlGroup
```

Reply:
I looked into the package and understand what you meant! I made the changes and added more unit tests to ensure that it is set.